thetis/limiter.py

```diff
@@ -86,19 +86,44 @@ def compute_bounds(self, field):
         """
         # Call general-purpose bound computation.
         super(VertexBasedP1DGLimiter, self).compute_bounds(field)
-        # NOTE This does not limit solution at lateral boundaries at all
-        # NOTE Omit for now
-        # # Add nodal values from lateral boundaries
-        # par_loop("""
-        #     for (int i=0; i<qmax.dofs; i++) {
-        #         qmax[i][0] = fmax(qmax[i][0], field[i][0]);
-        #         qmin[i][0] = fmin(qmin[i][0], field[i][0]);
-        #     }""",
-        #          ds,
-        #          {'qmax': (self.max_field, RW),
-        #           'qmin': (self.min_field, RW),
-        #           'field': (field, READ)})
 
+        # Add the average of lateral boundary facets to min/max fields
+        # NOTE this just computes the arithmetic mean of nodal values on the facet,
+        # which in general is not equivalent to the mean of the field over the bnd facet.
+        # This is OK for P1DG triangles, but not exact for the extruded case (quad facets)
+        from finat.finiteelementbase import entity_support_dofs
+
+        if self.is_2d:
+            entity_dim = 1  # get 1D facets
+        else:
+            entity_dim = (1, 1)  # get vertical facets
+        boundary_dofs = entity_support_dofs(self.P1DG.finat_element, entity_dim)
+        local_facet_nodes = np.array([boundary_dofs[e] for e in sorted(boundary_dofs.keys())])
+        n_bnd_nodes = local_facet_nodes.shape[1]
+        local_facet_idx = op2.Global(local_facet_nodes.shape, local_facet_nodes, dtype=np.int32, name='local_facet_idx')
+        code = """
+            void my_kernel(double **qmax, double **qmin, double **field, unsigned int *facet, unsigned int *local_facet_idx)
+            {
+                double face_mean = 0.0;
+                for (int i = 0; i < %(nnodes)d; i++) {
+                    unsigned int idx = local_facet_idx[facet[0]*%(nnodes)d + i];
```

Comment:
The indexing in here is a bit hacky... It wasn't obvious how to use a 2D int array in pyop2.